src/core/net/ip6.hpp

```diff
@@ -373,6 +374,7 @@ class Ip6 : public InstanceLocator, private NonCopyable
     Tasklet       mSendQueueTask;
 
     Icmp mIcmp;
+    Tcp  mTcp;
```

Comment:
I think it may be useful to add the `Tcp` under an OT config (e.g. `OPENTREAD_CONFIG_TCP6_ENABLE`). This way users have control whether to enable and use the OT TCP stack or not.

Reply:
I woud suggest `OPENTREAD_CONFIG_TCP_ENABLE` without `6`.

Reply:
I'm happy to add the flag. If the flag is off (set to `0`), then what should the API functions do?

Reply:
Is there a preferred error to return? And should that only be implemented in the external API, or in the internal API as well (i.e., if something in OpenThread uses `Tcp::Endpoint` and `Tcp::Listener` directly)?

Reply:
The convention so far has been to not implement the related functions when a given build feature is disabled, which would cause link failures with anything that tries to use them.

Whether we need to separate public from internal APIs depends on if there is a use case for it. For many features, we do not have such separation. However, for coaps we do - Thread Commissioning relies on coaps and applications may also want to leverage the same coaps implementation.

Reply:
Thanks Jonathan for clarifying. Since it's OK for applications that try to use TCP when it's disabled to fail to link, it sounds like we should also wrap most of `tcp6.cpp` in a `#ifdef` so that those functions aren't defined at all. Does that sound reasonable?

Reply:
Yes, that's what we do for other components. A couple examples:
https://github.com/openthread/openthread/blob/3a7e836dd7c96a932cbd8ab7733b1042bfd86588/src/core/api/coap_api.cpp#L36
https://github.com/openthread/openthread/blob/3a7e836dd7c96a932cbd8ab7733b1042bfd86588/src/core/api/jam_detection_api.cpp#L36